File: src/easy/string/ValidParentheses.java
package src.easy.string;

import java.util.LinkedList;

class ValidParentheses {
    public boolean isValid(String s) {
        // Start typing your Java solution below
        // DO NOT write main() function
        LinkedList<Character> stack = new LinkedList<>();
        for(int i=0; i<s.length(); i++){
            char ch = s.charAt(i);
            if(isLeftParenthese(ch))    stack.push(ch);
            else{
                if(!stack.isEmpty() && isPair(stack.peek(), ch))    stack.pop();
                else    return false;
            }
        }
        
        return stack.isEmpty();
    }
    
    private boolean isPair(char p, char q){
        if (p=='(' && q==')')   return true;
        if (p=='[' && q==']')   return true;
        if (p=='{' && q=='}')   return true;
        return false;
    }
    
    private boolean isLeftParenthese(char p){
        if (p=='(' || p=='[' || p=='{') return true;
        return false;
    }
}

File: src/python/string/valid_parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        map = {'(':')', '[':']', '{':'}'}
        q = []
        for c in s:
            if c in map:
                q.append(c)
            else:
                if q and map[q[-1]] == c:
                    q.pop()
                else:
                    return False
        return len(q)==0